File: frontend/src/app/shared/services/admin/evaluation.service.ts
import { HttpClient, HttpContext } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { ERROR_CONTEXT } from '@constant/error.registry';
import { environment } from '@env/environment';
import { IResponse } from '@model/response.model';
import {
  EvaluationBulkGenerateCommentRequest,
  EvaluationBulkSaveRequest,
  EvaluationEditWindowRequest,
  EvaluationEditWindowResponse,
  EvaluationGenerateCommentRequest,
  EvaluationSheetResponse,
} from '@app/model/admin/evaluation.model';

@Injectable({ providedIn: 'root' })
export class EvaluationService {
  private readonly baseUrl = `${environment.host_api}/evaluations`;
  private readonly silentContext = new HttpContext().set(ERROR_CONTEXT, {
    silent: true,
  });

  constructor(private readonly http: HttpClient) {}

  saveBulk(payload: EvaluationBulkSaveRequest) {
    return this.http.put<IResponse<unknown>>(`${this.baseUrl}/bulk`, payload, {
      context: this.silentContext,
    });
  }

  generateComment(payload: EvaluationGenerateCommentRequest) {
    return this.http.post<IResponse<string>>(`${this.baseUrl}/generate-comment`, payload, {
      context: this.silentContext,
    });
  }

  bulkGenerateComment(payload: EvaluationBulkGenerateCommentRequest) {
    return this.http.post<IResponse<{ [key: string]: string }>>(`${this.baseUrl}/bulk-generate-comment`, payload, {
      context: this.silentContext,
    });
  }

  getSheet(classroomId: string | number, subjectId: string | number, semesterId: string | number) {
    return this.http.get<IResponse<EvaluationSheetResponse>>(`${this.baseUrl}/sheet`, {
      params: { classroomId, subjectId, semesterId },
    });
  }

  getEditWindow(semesterId: string | number) {
    return this.http.get<IResponse<EvaluationEditWindowResponse | null>>(
      `${this.baseUrl}/edit-window`,
      {
        params: { semesterId },
        context: this.silentContext,
      }
    );
  }

  saveEditWindow(payload: EvaluationEditWindowRequest) {
    return this.http.put<IResponse<EvaluationEditWindowResponse>>(
      `${this.baseUrl}/edit-window`,
      payload,
      {
        context: this.silentContext,
      }
    );
  }

  exportTemplate(classroomId: number, subjectId: number, semesterId: number) {
    return this.http.get(`${this.baseUrl}/export-template`, {
      params: { classroomId, subjectId, semesterId },
      responseType: 'blob',
    });
  }

  importExcel(file: File, classroomId: number, subjectId: number, semesterId: number) {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('classroomId', classroomId.toString());
    formData.append('subjectId', subjectId.toString());
    formData.append('semesterId', semesterId.toString());

    return this.http.post<IResponse<any>>(`${this.baseUrl}/import`, formData);
  }

  downloadImportError(token: string) {
    return this.http.get(`${this.baseUrl}/import-errors/${token}`, {
      responseType: 'blob',
    });
  }
}

File: backend/src/main/java/com/gfi/backend/services/interfaces/EvaluationService.java
package com.gfi.backend.services.interfaces;

import com.gfi.backend.models.dtos.evaluation.EvaluationBulkUpsertRequest;
import com.gfi.backend.models.dtos.evaluation.EvaluationEditWindowDto;
import com.gfi.backend.models.dtos.evaluation.EvaluationEditWindowRequest;
import com.gfi.backend.models.dtos.evaluation.EvaluationGenerateCommentRequest;
import com.gfi.backend.models.dtos.evaluation.EvaluationSheetDto;

import com.gfi.backend.models.dtos.evaluation.EvaluationBulkGenerateCommentRequest;
import com.gfi.backend.models.dtos.common.TemporaryFileDto;
import com.gfi.backend.models.dtos.evaluation.EvaluationImportResultDto;
import org.springframework.web.multipart.MultipartFile;
import java.util.Map;

public interface EvaluationService {
    EvaluationSheetDto getSheet(Long classroomId, Long subjectId, Long semesterId);
    EvaluationEditWindowDto getEditWindow(Long semesterId);
    EvaluationEditWindowDto saveEditWindow(EvaluationEditWindowRequest request);
    void bulkUpsert(EvaluationBulkUpsertRequest request);
    String generateComment(EvaluationGenerateCommentRequest request);
    Map<Long, String> bulkGenerateComment(EvaluationBulkGenerateCommentRequest request);
    byte[] exportExcelTemplate(Long classroomId, Long subjectId, Long semesterId);
    EvaluationImportResultDto importExcel(MultipartFile file, Long classroomId, Long subjectId, Long semesterId);
    TemporaryFileDto getImportErrorFile(String token);
}
